rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,511 @@
+> Mitigates errors (3.3) and (3.4)
+
+External conditions such as failures from an unhealthy Elasticsearch cluster
+(3.3) can cause the migration to fail. The Kibana cluster should be able to
+recover automatically once these external conditions are resolved. There are
+two broad approaches to solving this problem based on whether or not
+migrations are idempotent: 
+
+| Idempotent migrations |Description                                                      |
+| --------------------- | --------------------------------------------------------------- |
+| Yes                   | 3.3.1 Idempotent migrations performed without coordination      |
+| No                    | 3.3.2 Single node migrations coordinated through a lease / lock |
+
+Because idempotent migrations don't require coordination, solution (3.3.1) is
+significantly less complex and it will never require manual intervention to
+retry. We, therefore, prefer this solution, even though it introduces new
+restrictions on migrations (4.3.1.1).
+
+## 4.3.1 Idempotent migrations performed without coordination
+
+The migration system can be said to be idempotent if the same results are
+produced whether the migration was run once or multiple times. This property
+should hold even if new (up to date) writes occur in between migration runs
+which introduces the following restrictions:
+
+### 4.3.1.1 Restrictions
+
+1. All document transforms need to be deterministic, that is a document
+   transform will always return the same result for the same set of inputs. 
+2. It should always be possible to construct the exact set of inputs required
+   for (1) at any point during the migration process (before, during, after). 
+
+### 4.3.1.2 Algorithm
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed. 
+2. Create `.kibana_n+1`, if it already exists, ignore the error and continue.
+3. Migrate documents by reading from `.kibana_n` and creating documents in
+   `.kibana_n+1`. If a document already exists, ignore.
+4. Move the `.kibana` index to point to `.kibana_n+1` (ignore if it has
+   already been moved). 
+
+Together with the limitations this algorithm ensures that migrations are
+idempotent. If two nodes are started simultaneously, both of them will start
+writing into `.kibana_n+1` but because migrations are idempotent, it doesn’t
+matter which node’s writes win.
+
+> Note: This algorithm doesn't support mixed Kibana versions attempting to
+> perform a migration at the same time. E.g. if a v7.8.0 and a v7.9.0 node
+> attempts to simultaneously migrate a 7.6.0 index the result will be an
+> inconsistent state. This is an unlikely scenario in practise and therefore
+> an acceptable limitation. In 8.x we can ensure that only the latest node
+> completes the migration.
+
+## 4.3.2 Single node migrations coordinated through a lease/lock
```

Comment:
This RFC should be making a strong proposal. If we're not planning to go down this route, I think this should moved to the "alternatives" section of the RFC along with a clear reason we are not opting for this option. I was a bit confused when first reading this and thought maybe this RFC was proposing we do both until I read the fine print at the top of 4.3

Reply:
Makes sense, I've moved it into the alternatives section.